### PR TITLE
QBitrain Strategy added

### PR DESCRIPTION
**Pull request by mablue**
```
@@@@@@@@@@@@@@@@&@#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@, @@/@@@&@@@@@@@@@&@*@@/@.@@(@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@%@*@@/%@&,@@#.@@@@@@#@@,.@@@/ @@ @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@*##@@ @.@@@&.@@@@@@@.@ &@@& .@@,@@&@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@&%@@% @@@  @@% @@@@@@@@@@**& @@@(#,@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@&,,. (@% . @@@@@@@@@@@( ,#.,%@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@&  @@@@@@@@@@@ (@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@   %@@@@@@( *@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@&/*/@@@@@@@@@@@@%             @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@%.,&@@%/@@@@@@@@@@                 /@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@( #@@@@@@@/                          ,      #@@@@@@@@@@@@@@@@@@@@@@@@@@
@%#*                                       %@@#          @@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@&. /@@@@@@&                     (@@@@@@@@@@@&.      @@@@@@@@@@@@@@@@@@@@@
@@@@*.&@ @@@@@@@@@%                   @@@@@@@@@@@@@@@@@*,    %@@@@@@@@@@@@@@@@@@
@@@@@@@(@@@@@@@@@                  *@@@@@@@@@@@@@@@@@@@@%     @@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@(   @@@@@@@@@@@@,   #@@@@@@@@@@@@@@@@@@@@@@@%    (@@@@@@@@@@@@@@@@@
@@@@@@@@(  @@@ / (@@@@@@@@@@@@@, &@@@@@@@@@@@@@@@@@@@@@@@#     @@@@@@@@@@@@@@@@@
@@@@@@*@@.@@@ @@ *&,@@@@@@@@@@@* ( %@@@@@@@@@@@@@@@@@@@@@#    .@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@ #@/&,@@#@@@@@@@( %@ (@@@,#@@@@@@@@@@@@@@@@@@@/ .@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@%%&&@#@#@@@@@@@@.@@@@%  ,@@@&%@#@@@@@@@@@@@@@@@     @@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@/@@@@* @&*@@.@@@%@@@@@@@@@@@@@@/     @@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@/@#@@,&@@@@@@@@@@@@@@@@@@@@&     &@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@&@@&@@@@@@@@@@@@@@@@@@     &@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@(   #@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@.    .@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@.     &@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@#     @@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@,    %@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@/   %@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@    .@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@     @@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@*    (@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@     @@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@# %@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@% %@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@.  @@@@@@@@% .#* ./,.*%@@%
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ &@@@%(/**.%@@@@@@@ &@@@@%
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@, &@@@@@@@@@@@ @@@@@@@(*@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@#  (@ &@@@@@@@@@@# @@@@@@@@  @@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@% @@@ @@@@,*@@@@@@@@@,@@@(  %@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ @@@@(*@@@@@@@# @@@@@@  @@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ .@@@@@&*@@@@@@@@@. &@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@  &@@@@@@@@@@@@@@@@@@@@@@@@@@@@
```
## QBitrain Strategy QBit + Brain + train
The idea is that some neurons of a brain connecting to each other to make a sum of
numbers that help to make decisions to buy and sell.Mixed With Quantum Bits State!
That multiple with the neuron value to change the neuron data to a wighted data.
-1 is fully reversed, 0 means fully disabled, 1 is fully active.
but its a range between -1 to 1 like quantum state of the particles!
Author: @Mablue (Masoud Azizi)
github: https://github.com/mablue/

### CODE REVIEW

**Review by xmatthias — CHANGES_REQUESTED** [review]
Please consider the "2 file" hyperopt interface as deprecated, and "to be removed" sometime in the near future.
The "parameter" interface allows the same functionalities, but in a much simpler way for the user (with a lot less doubts and manual effort).


As such - i'll not be accepting any PR's with explicit HO files anymore - and i've already started removing the existing Hyperopt files (migrating their logic to the respective strategies).

**Comment by mablue** [comment]
ok but you know why I am using 2file method.
for
hyperopting arrays
its necessary

**Comment by xmatthias** [comment]
Considering that your other strategies (heracles, ...) caused a lot of questions and doubts already, i don't think it's necessary, and will only complicate code that could be expressed in a better way, which makes people understand the strategy.

**Comment by mablue** [comment]
> Considering that your other strategies (heracles, ...) caused a lot of questions and doubts already, i don't think it's necessary, and will only complicate code that could be expressed in a better way, which makes people understand the strategy.

yes Them had bunch of problems and asking questions but why?
cuz its not mine idea!
heracles and Zeus and etc are the idea of computers and humans not familiar with CPUs. Im In shok as well. I dont know why people cant work with an small one indicator one file strategy.
but you will agree with that the biggest problem of data scientists is that them cant let everything to computers. I try to do that I want to force computer to write his own strategy line by line.I think its amazing but have custs. for example bunch of people that cant undrestand in first. I see them like who that allways have new technology fair.but when them try this new technology them allways will love it and use it. you know what I mean?
I mean our brain cant handle big data but CPUs If we struct them selfcontrolled and give everything in them hands, them can be useful. 
in other hand I know you care about complexity and yes Its a big problem. but its not the main code its just an strategy that works for his own in a side and many people love it I had more more messages that say me please continue and Improve. and why we dont do that If some people confused?!

**Comment by xmatthias** [comment]
I'm not against this concept in general - the only thing i asked is to do this using the new interface, as we'll probably be dropping support for the explicit hyperopt file later this year.

**Comment by mablue** [comment]
> I'm not against this concept in general - the only thing i asked is to do this using the new interface, as we'll probably be dropping support for the explicit hyperopt file later this year.
> 
> For that reason - having new explicit hyperopt files now is counter-productive, and would result in removal of this strategy/hyperopt pair in a few months.

I dont love old 2file method too.please remove it.
but I love being strong on writing strategys.
let find a way to do that.
I will try to do something that decrease your fair and attend you to do :))
but Yes now I seen your discord message and I know you didnt agree with it ... ok... ;)
I can find a way to do it without arrays

File: user_data/strategies/QBitrain.py [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+from pandas import DataFrame
+
+# --------------------------------
+
+# Add your lib to import here
+# TODO: talib is fast but have not more indicators
+import talib.abstract as ta
+from functools import reduce
+#  TODO: this gene is removed 'MAVP' cuz or error on periods
+AllLearntKnowledges = {
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:25]: Your variable usage is all over the place and makes it very hard to follow the strategy.

For a strategy with this complexity (which i expect to generate a lot of questions again) - please use common Python naming conventions - which will allow others with some python experience to easily identify what each variable corresponds to.

This means: 
* all uppercase for constants (things that are never modified again)
* camelcase for class names
* lowercase (seperated by `_` if it's multi-word) for variables.

File: user_data/strategies/QBitrain.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+
+        # TODO: Its not dry code!
+
+        conditions = []
+        RESULT = 0
+
+        IND = self.buy_indicator_0.value
+        DFINP = dataframe[IND] = knowledge_calculator(dataframe, IND)
+        QS = self.buy_node_quantum_state_0.value
+        RESULT += DFINP*QS
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:443]: Each of the below can be reduced to 1-2 lines (there's no need for `QS` - nor for `DFINP` (use `dataframe[IND]` for that.

Your way to express this is pretty verbose and makes the strategy appear more complicated than it really is.

File: user_data/strategies/QBitrain.py [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+        But this strategy can take much time to generate anything that may not use in his optimization.
+        I just calculate the specific indicators in specific time period inside buy and sell strategy populator methods if needed.
+        Also, this method (populate_indicators) just calculates default value of hyperoptable params
+        so using this method have not big benefits instade of calculating useable things inside buy and sell trand populators
+        '''
+        return dataframe
+
+    def populate_buy_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
+
+        # TODO: Its not dry code!
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:435]: These comments are pointless - if you think this is the case, then please reduce the verbosity (see below).

File: user_data/strategies/QBitrain.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+            return result
+        # For
+        elif knowledge_len == 3:
+            # print('knowledge_len == 3\t', indicator)
+            knowledge_timeperiod = int(knowledge[2])
+            knowledge_index = int(knowledge[1])
+            result = getattr(ta, knowledge_name)(
+                dataframe,
+                timeperiod=knowledge_timeperiod,
+            ).iloc[:, knowledge_index]
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:285]: Did you verify (for each and every ta-function that's listed in the list above) - that they REALLY have a `timeperiod` attribute?
there's some that don't have that IIRC - so i assume you made sure to not include these?

File: user_data/strategies/QBitrain.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+def normalize(df):
+    df = 2*(df-df.min())/(df.max()-df.min()) - 1
+    return df
+
+
+class QBitrain(IStrategy):
+    # #################### RESULTS LearntE PLACE ##################
+    # *   22/100:     16 trades. 13/3/0 Wins/Draws/Losses. Avg profit   4.92%. Median profit   4.31%. Total profit  0.26254039 BTC (  26.25Σ%). Avg duration 1 day, 14:00:00 min. Objective: -21.56636
+    # Buy hyperspace params:
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:320]: This result proves what i point out elsewhere - by having a stoploss as wide as this strategy uses, the strategy is based on luck / pairs eventually recovering - not on having learned a lot.

File: user_data/strategies/QBitrain.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+                timeperiod=knowledge_timeperiod,
+            ).iloc[:, knowledge_index]
+            return ta.SMA(dataframe[sharp_indicator].fillna(0), TREND_CHECK_CANDLES)
+
+# OK. Dont worry!! its used in end result
+
+
+def normalize(df):
+    df = 2*(df-df.min())/(df.max()-df.min()) - 1
+    return df
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:314]: This has a lookahead bias (as mentioned the last time, too).

Results created with this will not realize when using real modes that don't give you the full dataframe.

File: user_data/strategies/QBitrain.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+            f'{LearntKnowledge}-{timeperiod}')
+
+# Let give somethings to CatagoricalParam to Play with them
+# When just one thing is inside catagorical lists
+# TODO: its Not True Way :)
+if len(LearntKnowledges) == 1:
+    LearntKnowledges = LearntKnowledges*2
+if len(timeperiods) == 1:
+    timeperiods = timeperiods*2
+
```
Comment (xmatthias) [inline user_data/strategies/QBitrain.py:238]: it's unclear what you try to accomplish here (the comment above hints that you're also not sure).

why multiply things by 2 if the length is 1? that seems pretty much random ...

**Review by xmatthias — CHANGES_REQUESTED** [review]
Can you explain what makes this different to GodStrat(new) ? 
it seems like it's simply combining all indicators together to extract one value (probably a % of "good or bad?) ... 

Considering that we've been in a bull market for the last 2-3 years (for the most part) - you'll always make profits with a stoploss this wide.
I doubt you'll be able to make good profits when you reduce the stoploss to something below 10% (and remove the lookahead bias) - as this will probably result in a strategy that buys based on one indicator-group - where all basically tell the same story in slightly different terms - which will reinforce each other - without really having learned something.


One thing i'll want to have fixed for sure is the messy variable naming. Otherwise, i'll not be able to accept this (accepting a PR means also taking responsibility to eventually maintain something - which with this messy code, i currently cannot).
